### PR TITLE
Create a fix to support requested encoding type in next_key

### PR DESCRIPTION
**Pull request by nickjjzhao**
## Change Description
The fix in this PR is to make sure that the value returned in "`next_key`" matches the `--encode-type` of the request that caused it to be returned (and thus can be used to replace the original passed in `--lower`, or `--upper` with `--reverse`, parameter and the request repeated).

## Change Type
**Select *ONE*:**
- [ ] Documentation

- [x] Stability bug fix

- [ ] Other

- [ ] Other - special case

## Testing Changes
**Select *ANY* that apply:**
- [ ] New Tests

- [x] Existing Tests

- [ ] Test Framework

- [ ] CI System

- [ ] Other

## Consensus Changes
- [ ] Consensus Changes

## API Changes
- [ ] API Changes

## Documentation Additions
- [ ] Documentation Additions

### CODE REVIEW

**Comment by brianjohnson5972** [comment]
@nickjjzhao Please verify that the my edit to the initial description is correct.  I'm guessing that you had made the PR via the hub tool.

**Comment by nickjjzhao** [comment]
> @nickjjzhao Please verify that the my edit to the initial description is correct. I'm guessing that you had made the PR via the hub tool.
Thanks Brian! Your edit is correct.

File: plugins/chain_plugin/chain_plugin.cpp [hunk 1]
```diff
@@ -2408,9 +2408,164 @@ read_only::get_table_rows_result read_only::get_kv_table_rows_context( const rea
...
+            std::memcpy(&t, &val, sizeof(T));
+        }else {
+            static_assert(sizeof(T) == 8, "Unknown floating point type");
+            uint64_t val = float_from_bytes<uint64_t>(t, next_key_bytes);
+            std::memcpy(&t, &val, sizeof(T));
+        }
+    }else if constexpr (std::is_integral_v<T>) {
+        std::stringstream ss;
+        ss << std::hex << next_key_bytes;
+        auto unsigned_val = static_cast<std::make_unsigned_t<T>>(t);
```
Comment (kimjh2005) [inline plugins/chain_plugin/chain_plugin.cpp:2445]: "t" is not initialized.

File: tests/get_kv_table_nodeos_tests.cpp [hunk 2]
```diff
@@ -796,6 +955,31 @@ BOOST_FIXTURE_TEST_CASE( get_kv_table_nodeos_test, TESTER ) try {
    chk_result(0, 4);
    chk_result(1, 5);
 
+   // test next_key
+   p.reverse = false;
+   p.index_name = "ff"_n;
+   p.index_value = "";
+   p.encode_type = "dec";
+   p.lower_bound = "0.02";
```
Comment (kimjh2005) [inline tests/get_kv_table_nodeos_tests.cpp:963]: Can you add more test cases including negative float numbers?

Comment (nickjjzhao) [reply tests/get_kv_table_nodeos_tests.cpp:985]: Fixed.

File: tests/get_kv_table_nodeos_tests.cpp [hunk 3]
```diff
@@ -796,6 +955,53 @@ BOOST_FIXTURE_TEST_CASE( get_kv_table_nodeos_test, TESTER ) try {
...
+   result = plugin.read_only::get_kv_table_rows(p);
+   BOOST_REQUIRE_EQUAL(0u, result.rows.size());
+   BOOST_REQUIRE_EQUAL(result.more, false);
+
+   p.upper_bound = "";
+   p.limit = 1;
+   result = plugin.read_only::get_kv_table_rows(p);
+   BOOST_REQUIRE_EQUAL(1u, result.rows.size());
+   BOOST_REQUIRE_EQUAL(result.more, true);
+   BOOST_REQUIRE_EQUAL(result.next_key != "", true);
```
Comment (kimjh2005) [inline tests/get_kv_table_nodeos_tests.cpp:975]: Can you use real value  instead of   next_key != ""  ?
   ex)  next_key == "some real value"

Comment (nickjjzhao) [reply tests/get_kv_table_nodeos_tests.cpp:975]: I tried using "some real value" for the following checks. One check failed and the other passed.
BOOST_REQUIRE_EQUAL(result.next_key == "1.01000000000000001", true); - Failed
BOOST_REQUIRE_EQUAL(result.next_key == "1.010000", true); - Passed

Is there a way to control the precision of returned result.next_key, which is decoded from result.next_key_bytes?

Comment (kimjh2005) [reply tests/get_kv_table_nodeos_tests.cpp:975]: I think you can just check  a sub string of "1.01"

Comment (nickjjzhao) [reply tests/get_kv_table_nodeos_tests.cpp:975]: Fixed.

**Review by kimjh2005 — APPROVED** [review]

File: tests/get_kv_table_nodeos_tests.cpp [hunk 4]
```diff
@@ -669,6 +803,31 @@ BOOST_FIXTURE_TEST_CASE( get_kv_table_nodeos_test, TESTER ) try {
...
+   p.index_name = "ii"_n;
+   p.encode_type = "dec";
+   p.index_value = "";
+   p.lower_bound = "-1000";
+   p.upper_bound = "1000";
+   p.limit = 2;
+   result = plugin.read_only::get_kv_table_rows(p);
+   BOOST_REQUIRE_EQUAL(2u, result.rows.size());
+   BOOST_REQUIRE_EQUAL(result.more, true);
+   BOOST_REQUIRE_EQUAL(result.next_key != "", true);
```
Comment (brianjohnson5972) [inline tests/get_kv_table_nodeos_tests.cpp:817]: check actual value

File: tests/get_kv_table_nodeos_tests.cpp [hunk 5]
```diff
@@ -669,6 +803,31 @@ BOOST_FIXTURE_TEST_CASE( get_kv_table_nodeos_test, TESTER ) try {
...
+   chk_result(0, 1);
+   chk_result(1, 2);
+
+   p.lower_bound = result.next_key;
+   p.upper_bound = "";
+   p.limit = 2;
+   result = plugin.read_only::get_kv_table_rows(p);
+   BOOST_REQUIRE_EQUAL(2u, result.rows.size());
+   BOOST_REQUIRE_EQUAL(result.more, true);
+   BOOST_REQUIRE_EQUAL(result.next_key != "", true);
```
Comment (brianjohnson5972) [inline tests/get_kv_table_nodeos_tests.cpp:827]: check actual value

File: tests/get_kv_table_nodeos_tests.cpp [hunk 6]
```diff
@@ -796,6 +955,31 @@ BOOST_FIXTURE_TEST_CASE( get_kv_table_nodeos_test, TESTER ) try {
...
+   p.index_name = "ff"_n;
+   p.index_value = "";
+   p.encode_type = "dec";
+   p.lower_bound = "0.02";
+   p.upper_bound = "3.03000001";
+   p.limit = 2;
+   result = plugin.read_only::get_kv_table_rows(p);
+   BOOST_REQUIRE_EQUAL(2u, result.rows.size());
+   BOOST_REQUIRE_EQUAL(result.more, true);
+   BOOST_REQUIRE_EQUAL(result.next_key != "", true);
```
Comment (brianjohnson5972) [inline tests/get_kv_table_nodeos_tests.cpp:995]: check actual value

File: tests/get_kv_table_nodeos_tests.cpp [hunk 7]
```diff
@@ -796,6 +955,31 @@ BOOST_FIXTURE_TEST_CASE( get_kv_table_nodeos_test, TESTER ) try {
...
+   chk_result(0, 2);
+   chk_result(1, 3);
+
+   p.lower_bound = result.next_key;
+   p.upper_bound = "";
+   p.limit = 2;
+   result = plugin.read_only::get_kv_table_rows(p);
+   BOOST_REQUIRE_EQUAL(2u, result.rows.size());
+   BOOST_REQUIRE_EQUAL(result.more, true);
+   BOOST_REQUIRE_EQUAL(result.next_key != "", true);
```
Comment (brianjohnson5972) [inline tests/get_kv_table_nodeos_tests.cpp:1005]: check actual value

File: tests/get_kv_table_nodeos_tests.cpp [hunk 8]
```diff
@@ -230,6 +230,45 @@ BOOST_FIXTURE_TEST_CASE( get_kv_table_nodeos_test, TESTER ) try {
    chk_result(3, 9);
    chk_result(4, 10);
 
+   // test next_key
```
Comment (brianjohnson5972) [inline tests/get_kv_table_nodeos_tests.cpp:233]: all BOOST_REQUIRE_EQUAL(result.next_key_bytes != "", true); checks should be changed to verifying the expected behavior
I know this isn't your fault, but the test should be verifying the value, and we should be verifying the result.next_key value as well. If any of the cases you added are repeats of existing cases then we should remove them and just add the result.next_key check in the existing one. But make sure that we don't lose the checks I added above for boundary conditions like "dec" being "10" and "hex" being "a"(or "A") when the uinXX_t has a value of 0x0a.

For any of your new cases that stays around, also add verifying the result.next_key_bytes value as well.

**Review by brianjohnson5972 — CHANGES_REQUESTED** [review]

File: plugins/chain_plugin/chain_plugin.cpp [hunk 9]
```diff
@@ -2408,9 +2408,164 @@ read_only::get_table_rows_result read_only::get_kv_table_rows_context( const rea
...
+            std::string s = result.next_key_bytes;
+            std::stringstream ss;
+            ss << std::hex << s;
+            uint64_t ull;
+            ss >> ull;
+            name nm(ull);
+            result.next_key = nm.to_string();
+        }else if (encode_type == "hex") {
+            if (index_type == "uint64") {
+                uint64_t u64;
```
[the remark this answers is not in the document]
Comment (nickjjzhao) [reply plugins/chain_plugin/chain_plugin.cpp]: Fixed

File: plugins/chain_plugin/chain_plugin.cpp [hunk 10]
```diff
@@ -2408,9 +2408,164 @@ read_only::get_table_rows_result read_only::get_kv_table_rows_context( const rea
...
+
+void read_only::set_kv_next_key(const string& encode_type, const string& index_type, read_only::get_table_rows_result& result) const {
+    if (result.more == true) {
+        if (encode_type == "bytes") {
+            result.next_key = result.next_key_bytes;
+        }else if (encode_type == "string") {
+            result.next_key = boost::algorithm::unhex(result.next_key_bytes);
+            boost::algorithm::trim_right_if( result.next_key, []( char c ){ return c == '\0'; } );
+        }else if (encode_type == "name") {
+            std::string s = result.next_key_bytes;
```
[the remark this answers is not in the document]
Comment (nickjjzhao) [reply plugins/chain_plugin/chain_plugin.cpp]: Fixed.

File: plugins/chain_plugin/chain_plugin.cpp [hunk 11]
```diff
@@ -2408,9 +2408,164 @@ read_only::get_table_rows_result read_only::get_kv_table_rows_context( const rea
...
+                result.next_key = result.next_key_bytes;
+            } else if( index_type == "ripemd160" ) {
+                result.next_key = result.next_key_bytes;
+            }else {
+                EOS_ASSERT(false, chain::contract_table_query_exception, "Unsupported index type/encode_type: ${t}/${e} ", ("t", index_type)("e", encode_type));
+            }
+
+        }else if (encode_type == "dec") {
+            if (index_type == "float64") {
+                double d;
```
[the remark this answers is not in the document]
Comment (nickjjzhao) [reply plugins/chain_plugin/chain_plugin.cpp]: Fixed.

File: plugins/chain_plugin/chain_plugin.cpp [hunk 12]
```diff
@@ -2408,9 +2408,164 @@ read_only::get_table_rows_result read_only::get_kv_table_rows_context( const rea
...
+            }else if (index_type == "int8") {
+                int8_t i8;
+                convert_from_bytes(i8, result.next_key_bytes);
+                convert_to_hex(i8, result.next_key);
+            }else if( index_type == "sha256" || index_type == "i256" ) {
+                result.next_key = result.next_key_bytes;
+            } else if( index_type == "ripemd160" ) {
+                result.next_key = result.next_key_bytes;
+            }else {
+                EOS_ASSERT(false, chain::contract_table_query_exception, "Unsupported index type/encode_type: ${t}/${e} ", ("t", index_type)("e", encode_type));
```
[the remark this answers is not in the document]
Comment (nickjjzhao) [reply plugins/chain_plugin/chain_plugin.cpp]: Fixed.

File: plugins/chain_plugin/chain_plugin.cpp [hunk 13]
```diff
@@ -2408,9 +2408,164 @@ read_only::get_table_rows_result read_only::get_kv_table_rows_context( const rea
...
+            }else if (index_type == "int16") {
+                int16_t i16;
+                convert_from_bytes(i16, result.next_key_bytes);
+                result.next_key = std::to_string(i16);
+            }else if (index_type == "int8") {
+                int8_t i8;
+                convert_from_bytes(i8, result.next_key_bytes);
+                result.next_key = std::to_string(i8);
+            }else {
+                EOS_ASSERT(false, chain::contract_table_query_exception, "Unsupported index type/encode_type: ${t}/${e} ", ("t", index_type)("e", encode_type));
```
[the remark this answers is not in the document]
Comment (nickjjzhao) [reply plugins/chain_plugin/chain_plugin.cpp]: Fixed. What's the major difference here by using EOS_ASSERT or FC_THROW_EXCEPTION?

File: tests/get_kv_table_nodeos_tests.cpp [hunk 14]
```diff
@@ -230,6 +230,45 @@ BOOST_FIXTURE_TEST_CASE( get_kv_table_nodeos_test, TESTER ) try {
...
+   p.index_value = "";
+   p.reverse = false;
+   p.encode_type = "name";
+   p.lower_bound = "boba";
+   p.upper_bound = "";
+   p.limit = 2;
+   result = plugin.read_only::get_kv_table_rows(p);
+   BOOST_REQUIRE_EQUAL(2u, result.rows.size());
+   BOOST_REQUIRE_EQUAL(result.more, true);
+   BOOST_REQUIRE_EQUAL(result.next_key != "", true);
```
[the remark this answers is not in the document]
Comment (nickjjzhao) [reply tests/get_kv_table_nodeos_tests.cpp]: Fixed.

File: tests/get_kv_table_nodeos_tests.cpp [hunk 15]
```diff
@@ -230,6 +230,45 @@ BOOST_FIXTURE_TEST_CASE( get_kv_table_nodeos_test, TESTER ) try {
...
+   chk_result(0, 1);
+   chk_result(1, 2);
+
+   p.lower_bound = result.next_key;
+   result = plugin.read_only::get_kv_table_rows(p);
+   BOOST_REQUIRE_EQUAL(2u, result.rows.size());
+   chk_result(0, 3);
```
[the remark this answers is not in the document]
Comment (nickjjzhao) [reply tests/get_kv_table_nodeos_tests.cpp:252]: Fixed.

File: tests/get_kv_table_nodeos_tests.cpp [hunk 16]
```diff
@@ -230,6 +230,45 @@ BOOST_FIXTURE_TEST_CASE( get_kv_table_nodeos_test, TESTER ) try {
...
+   chk_result(1, 4);
+
+   p.lower_bound = result.next_key;
+   p.limit = 1;
+   result = plugin.read_only::get_kv_table_rows(p);
+   BOOST_REQUIRE_EQUAL(1u, result.rows.size());
+   BOOST_REQUIRE_EQUAL(result.more, true);
+   chk_result(0, 5);
```
[the remark this answers is not in the document]
Comment (nickjjzhao) [reply tests/get_kv_table_nodeos_tests.cpp:261]: Fixed.

File: tests/get_kv_table_nodeos_tests.cpp [hunk 17]
```diff
@@ -230,6 +230,45 @@ BOOST_FIXTURE_TEST_CASE( get_kv_table_nodeos_test, TESTER ) try {
...
+
+   p.lower_bound = result.next_key;
+   p.limit = 20;
+   result = plugin.read_only::get_kv_table_rows(p);
+   BOOST_REQUIRE_EQUAL(5u, result.rows.size());
+   BOOST_REQUIRE_EQUAL(result.more, false);
+   chk_result(0, 6);
```
[the remark this answers is not in the document]
Comment (nickjjzhao) [reply tests/get_kv_table_nodeos_tests.cpp:269]: Fixed.

File: tests/get_kv_table_nodeos_tests.cpp [hunk 18]
```diff
@@ -357,6 +396,42 @@ BOOST_FIXTURE_TEST_CASE( get_kv_table_nodeos_test, TESTER ) try {
...
+   p.index_value = "";
+   p.encode_type = "dec";
+
+   p.lower_bound = "0";
+   p.upper_bound = "";
+   p.limit = 2;
+   result = plugin.read_only::get_kv_table_rows(p);
+   BOOST_REQUIRE_EQUAL(2u, result.rows.size());
+   BOOST_REQUIRE_EQUAL(result.more, true);
+   chk_result(0, 1);
```
[the remark this answers is not in the document]
Comment (nickjjzhao) [reply tests/get_kv_table_nodeos_tests.cpp:448]: Fixed.

File: tests/get_kv_table_nodeos_tests.cpp [hunk 19]
```diff
@@ -357,6 +396,42 @@ BOOST_FIXTURE_TEST_CASE( get_kv_table_nodeos_test, TESTER ) try {
...
+   chk_result(1, 2);
+
+   p.lower_bound = result.next_key;
+   p.limit = 3;
+   result = plugin.read_only::get_kv_table_rows(p);
+   BOOST_REQUIRE_EQUAL(3u, result.rows.size());
+   BOOST_REQUIRE_EQUAL(result.more, true);
+   chk_result(0, 3);
```
[the remark this answers is not in the document]
Comment (nickjjzhao) [reply tests/get_kv_table_nodeos_tests.cpp:457]: Fixed.

File: tests/get_kv_table_nodeos_tests.cpp [hunk 20]
```diff
@@ -357,6 +396,42 @@ BOOST_FIXTURE_TEST_CASE( get_kv_table_nodeos_test, TESTER ) try {
...
+   chk_result(1, 4);
+   chk_result(2, 5);
+
+   p.lower_bound = result.next_key;
+   p.limit = 20;
+   result = plugin.read_only::get_kv_table_rows(p);
+   BOOST_REQUIRE_EQUAL(5, result.rows.size());
+   BOOST_REQUIRE_EQUAL(result.more, false);
+   chk_result(0, 6);
```
[the remark this answers is not in the document]
Comment (nickjjzhao) [reply tests/get_kv_table_nodeos_tests.cpp:467]: Fixed.

File: tests/get_kv_table_nodeos_tests.cpp [hunk 21]
```diff
@@ -357,6 +396,42 @@ BOOST_FIXTURE_TEST_CASE( get_kv_table_nodeos_test, TESTER ) try {
...
+   p.limit = 3;
+   result = plugin.read_only::get_kv_table_rows(p);
+   BOOST_REQUIRE_EQUAL(3u, result.rows.size());
+   BOOST_REQUIRE_EQUAL(result.more, true);
+   chk_result(0, 3);
+   chk_result(1, 4);
+   chk_result(2, 5);
+
+   p.lower_bound = result.next_key;
+   p.limit = 20;
```
[the remark this answers is not in the document]
Comment (nickjjzhao) [reply tests/get_kv_table_nodeos_tests.cpp]: Fixed.

File: tests/get_kv_table_nodeos_tests.cpp [hunk 22]
```diff
@@ -514,6 +589,40 @@ BOOST_FIXTURE_TEST_CASE( get_kv_table_nodeos_test, TESTER ) try {
...
+   p.encode_type = "string";
+   p.reverse = false;
+   p.index_value = "";
+   p.lower_bound = "boba";
+   p.upper_bound = "";
+   p.limit = 2;
+   result = plugin.read_only::get_kv_table_rows(p);
+   BOOST_REQUIRE_EQUAL(2u, result.rows.size());
+   BOOST_REQUIRE_EQUAL(result.more, true);
+   chk_result(0, 1);
```
[the remark this answers is not in the document]
Comment (nickjjzhao) [reply tests/get_kv_table_nodeos_tests.cpp:687]: Fixed.

File: tests/get_kv_table_nodeos_tests.cpp [hunk 23]
```diff
@@ -357,6 +396,42 @@ BOOST_FIXTURE_TEST_CASE( get_kv_table_nodeos_test, TESTER ) try {
...
+   result = plugin.read_only::get_kv_table_rows(p);
+   BOOST_REQUIRE_EQUAL(5, result.rows.size());
+   BOOST_REQUIRE_EQUAL(result.more, false);
+   chk_result(0, 6);
+   chk_result(1, 7);
+   chk_result(2, 8);
+   chk_result(3, 9);
+   chk_result(4, 10);
+
+
```
[the remark this answers is not in the document]
Comment (nickjjzhao) [reply tests/get_kv_table_nodeos_tests.cpp:517]: Fixed.

File: tests/get_kv_table_nodeos_tests.cpp [hunk 24]
```diff
@@ -230,6 +230,45 @@ BOOST_FIXTURE_TEST_CASE( get_kv_table_nodeos_test, TESTER ) try {
...
+   p.limit = 20;
+   result = plugin.read_only::get_kv_table_rows(p);
+   BOOST_REQUIRE_EQUAL(5u, result.rows.size());
+   BOOST_REQUIRE_EQUAL(result.more, false);
+   chk_result(0, 6);
+   chk_result(1, 7);
+   chk_result(2, 8);
+   chk_result(3, 9);
+   chk_result(4, 10);
+
```
[the remark this answers is not in the document]
Comment (nickjjzhao) [reply tests/get_kv_table_nodeos_tests.cpp:274]: Fixed.

File: tests/get_kv_table_nodeos_tests.cpp [hunk 25]
```diff
@@ -609,6 +718,31 @@ BOOST_FIXTURE_TEST_CASE( get_kv_table_nodeos_test, TESTER ) try {
...
+   p.index_name = "i"_n;
+   p.index_value = "";
+   p.encode_type = "dec";
+   p.lower_bound = "-100";
+   p.upper_bound = "100";
+   p.limit = 2;
+   result = plugin.read_only::get_kv_table_rows(p);
+   BOOST_REQUIRE_EQUAL(2u, result.rows.size());
+   BOOST_REQUIRE_EQUAL(result.more, true);
+   BOOST_REQUIRE_EQUAL(result.next_key != "", true);
```
[the remark this answers is not in the document]
Comment (nickjjzhao) [reply tests/get_kv_table_nodeos_tests.cpp]: Fixed.

File: tests/get_kv_table_nodeos_tests.cpp [hunk 26]
```diff
@@ -609,6 +718,31 @@ BOOST_FIXTURE_TEST_CASE( get_kv_table_nodeos_test, TESTER ) try {
...
+   chk_result(0, 1);
+   chk_result(1, 2);
+
+   p.lower_bound = result.next_key;
+   p.upper_bound = "";
+   p.limit = 2;
+   result = plugin.read_only::get_kv_table_rows(p);
+   BOOST_REQUIRE_EQUAL(2u, result.rows.size());
+   BOOST_REQUIRE_EQUAL(result.more, true);
+   BOOST_REQUIRE_EQUAL(result.next_key != "", true);
```
[the remark this answers is not in the document]
Comment (nickjjzhao) [reply tests/get_kv_table_nodeos_tests.cpp]: Fixed.

File: tests/get_kv_table_nodeos_tests.cpp (hunk 4 above)
Comment (nickjjzhao) [reply tests/get_kv_table_nodeos_tests.cpp]: Fixed.

File: tests/get_kv_table_nodeos_tests.cpp (hunk 5 above)
Comment (nickjjzhao) [reply tests/get_kv_table_nodeos_tests.cpp]: Fixed.

File: tests/get_kv_table_nodeos_tests.cpp (hunk 6 above)
Comment (nickjjzhao) [reply tests/get_kv_table_nodeos_tests.cpp]: Fixed.

File: tests/get_kv_table_nodeos_tests.cpp (hunk 7 above)
Comment (nickjjzhao) [reply tests/get_kv_table_nodeos_tests.cpp]: Fixed.

File: tests/get_kv_table_nodeos_tests.cpp (hunk 8 above)
Comment (nickjjzhao) [reply tests/get_kv_table_nodeos_tests.cpp:233]: > For any of your new cases that stays around, also add verifying the result.next_key_bytes value as well.

Should we verify the actual value of result.next_key_bytes or verify its emptiness? It seems to me that checking actual value doesn't help much because a value of bytes type is an encoded hex.

File: plugins/chain_plugin/chain_plugin.cpp (hunk 1 above)
Comment (brianjohnson5972) [reply plugins/chain_plugin/chain_plugin.cpp]: no, either it will get set, or we will get an exception

File: plugins/chain_plugin/chain_plugin.cpp [hunk 27]
```diff
@@ -2408,9 +2408,136 @@ read_only::get_table_rows_result read_only::get_kv_table_rows_context( const rea
...
+            auto unsigned_val = static_cast<std::make_unsigned_t<T>>(t);
+            ss >> unsigned_val;
+            if (unsigned_val > std::numeric_limits<T>::max()) {
+                t = unsigned_val + static_cast<std::make_unsigned_t<T>>(std::numeric_limits<T>::min());
+            }
+            else {
+                t = unsigned_val + std::numeric_limits<T>::min();
+            }
+        }else {
+            EOS_ASSERT(false, chain::contract_table_query_exception, "Unsupported type to convert from bytes");
```
Comment (brianjohnson5972) [inline plugins/chain_plugin/chain_plugin.cpp:2454]: make FC_THROW_EXCEPTION

File: plugins/chain_plugin/chain_plugin.cpp (hunk 1 above)
Comment (nickjjzhao) [reply plugins/chain_plugin/chain_plugin.cpp]: @brianjohnson5972, do you mean "t" should be initialized? I didn't see any exception without initializing it.

File: tests/get_kv_table_nodeos_tests.cpp (hunk 8 above)
Comment (nickjjzhao) [reply tests/get_kv_table_nodeos_tests.cpp:233]: > For any of your new cases that stays around, also add verifying the result.next_key_bytes value as well.

Fixed.

File: tests/get_kv_table_nodeos_tests.cpp [hunk 28]
```diff
@@ -796,6 +1040,55 @@ BOOST_FIXTURE_TEST_CASE( get_kv_table_nodeos_test, TESTER ) try {
...
+   result = plugin.read_only::get_kv_table_rows(p);
+   BOOST_REQUIRE_EQUAL(0u, result.rows.size());
+   BOOST_REQUIRE_EQUAL(result.more, false);
+
+   p.upper_bound = "";
+   p.limit = 2;
+   result = plugin.read_only::get_kv_table_rows(p);
+   BOOST_REQUIRE_EQUAL(2u, result.rows.size());
+   BOOST_REQUIRE_EQUAL(result.more, true);
+   BOOST_REQUIRE_EQUAL(result.next_key.substr(0, 4) == "2.02", true);
```
Comment (brianjohnson5972) [inline tests/get_kv_table_nodeos_tests.cpp:1086]: change all these to be .find("2.02") != std::npos, to avoid this failing with an exception if for some reason next_key is less than 4 characters long

**Review by brianjohnson5972 — CHANGES_REQUESTED** [review]

File: plugins/chain_plugin/chain_plugin.cpp (hunk 27 above)
Comment (nickjjzhao) [reply plugins/chain_plugin/chain_plugin.cpp]: Fixed.

File: tests/get_kv_table_nodeos_tests.cpp (hunk 28 above)
Comment (nickjjzhao) [reply tests/get_kv_table_nodeos_tests.cpp]: Fixed.

**Review by brianjohnson5972 — APPROVED** [review]
